Add commands to create training materials

bebatut opened the pull request:
Hi,

This PR adds new commands to planemo (as discussed during GCCBOSC: https://github.com/galaxyproject/training-material/issues/941):

1. `training_init`: a command to create new training material for the GTN material given the template there:
   1. It creates a topic if not present (with metadata filled)
   2. It creates a tutorial if not present (with all needed files)
   3. It generates the skeleton of a tutorial if a workflow is given and add also the workflow
   4. It creates the `data_library.yaml` file by parsing the Zenodo API and add the link in the tutorial
   5. It updates the metadata if needed
2. `training_generate_tuto_from_wf` to generate the `tutorial.md` of an existing tutorial from a workflow
3. `training_fill_data_library` to fill the `data_library.yaml` file by parsing the Zenodo API of an existing tutorial

I tested it by generating several tutorial skeletons. @abretaud tested it too.

I added tests to cover most of the functions (I have some doubts about the tests of the commands).
I am not sure about the code quality, particularly the `training.py` script if not so straightforward. 
I hope I could get your help there :smile:

ping @shiltemann, @jmchilton 

Bérénice

bebatut commented:
😟 
Can you give me your workflow?

abretaud commented:
It's this one: https://github.com/abretaud/training-material/blob/d5c75ade7ef436b6fc93dd40cf5fa90e3a939dbf/topics/genome-annotation/tutorials/annotation-with-maker/workflows/init_workflow.ga

bebatut commented:
@abretaud it should be fixed. can you test? thanks

abretaud commented:
@bebatut I confirm! thanks

abretaud commented:
I've just re-checked the tutorial.md generated with the latest code on the [maker workflow](https://github.com/abretaud/training-material/blob/d5c75ade7ef436b6fc93dd40cf5fa90e3a939dbf/topics/genome-annotation/tutorials/annotation-with-maker/workflows/init_workflow.ga):
- some select boxes are still absent from the tutorial: e.g. "Organism type" for Maker tool 
- the same for a multiple select box: "Select fasta outputs" for gffread
The rest looks good :)

jmchilton commented:
Some of the tests just don't pass for me:

```python
    n = 'save_db'
    tp_desc = tools[step['name']][n]
    step_param = training.get_lower_params(step_params, n)
    print(tp_desc)
    print(n)
    print(step_params)
    print(step_param)
    paramlist = training.format_param_desc(
        step_param,
        step_inputs,
        tp_desc,
        0,
        wf['steps'],
        force_default=False)
    assert '`Yes`' in paramlist, paramlist
```

That parameter appears to be false not true for instance. Does that work for you locally - does it require Galaxy modifications?

abretaud commented on project_templates/training/tutorials/tutorial1/data-library.yaml:
```diff
@@ -0,0 +1,7 @@
+---
```
Is this file still needed? It looks its format changed

abretaud commented on planemo/training.py:
```diff
@@ -0,0 +1,938 @@
+>    > ### {{ '{%' }} icon tip {{ '%}' }} Tip: Importing data from a data library
+>    >
+>    > * Go into "Shared data" (top panel) then "Data libraries"
+>    > * Click on "Training data" and then "{{ topic_title }}"
```
Is it only `topic_title` or `topic_title`/`tutorial_title`?

abretaud commented on planemo/training.py:
```diff
@@ -0,0 +1,938 @@
+        wf, tools = get_wf_tools_from_running_galaxy(kwds)
+    else:
+        wf, tools = get_wf_tool_from_local_galaxy(kwds, kwds["workflow"], ctx)
+    save_to_yaml(tools, 'tools.yaml')
```
Hum, don't know if it's intended but now I get a tools.yaml at the root of the gtn dir (but it doesn't look the same as the tutorial tools.yaml)

bebatut commented:
I updated to fit with the new metadata definition

@jmchilton The test should be fixed now
@abretaud I fixed some of your comments. I need to test again your workflow to investigate the issue with the boxes... Next week

bebatut commented:
@abretaud I checked your comments. If the value of a parameter is set to the default value in the tool description, the corresponding parameter description does not appear (to avoid all parameters to be described, specially for big tool). It explain why "Organism type" for Maker tool does not appear
Does it make sense?

abretaud commented:
@bebatut ah yes, it makes sense, I thought it was not the default value

bebatut commented:
I restructured the code and split it into classes to try to make it easier to maintain.
I extracted the doc for it.
I think it is ready for review. ping @jmchilton, @shiltemann, @erasche

mblue9 commented:
Hi @bebatut !

I'm trying `training_generate_from_wf` out at the moment and getting an error `TypeError: string indices must be integers`. This is the workflow I'm using (I just changed the file ending to .txt from .ga to add it here)
[Galaxy-Workflow-limma-voom_tutorial_fastqs_to_counts.txt](https://github.com/galaxyproject/planemo/files/2365359/Galaxy-Workflow-limma-voom_tutorial_fastqs_to_counts.txt)

Btw I'm using python 2.7 with this but I see python 3.6 in @abretaud's error above, should I be using python 3 or it doesn't matter?

bebatut commented:
@mblue9 I could not reproduce your error (with Python 2.7 or 3.6) 😟
Did you use the latest version from this PR? What command did you use?

mblue9 commented:
@bebatut and @jmchilton thanks for taking a look! Yes I used the version from this PR and the command I used was: `planemo training_generate_from_wf --topic_name transcriptomics --tutorial_name "limma-voom_fastqs_to_counts"  --workflow ~/training/pmac/galaxy/limma-voom/limma-voom_fastqs_to_counts/Galaxy-Workflow-limma-voom_tutorial_fastqs_to_counts.ga --galaxy_root ~/galaxy/galaxy/`

(The `training_init` command seemed to work)

@bebatut does it work for you with my workflow? If so do you think you could send me the tutorial.md and I could keep going with that? as I'm running this tutorial in a couple of weeks. Otherwise I'll just manually create it.

bebatut commented:
@jmchilton All the tests implying to start a Galaxy instance (though the call of `get_hands_on_boxes_from_local_galaxy` or in which I started a Galaxy instance) and then getting the workflow and the tools from this Galaxy instancne seems to fail on Travis (but locally for me):

```
Test :func:`planemo.training.generate_tuto_from_wf`. ... ERROR
Test :func:`planemo.training.tutorial.format_wf_steps`. ... ERROR
Test :func:`planemo.training.tutorial.get_hands_on_boxes_from_local_galaxy`. ... ERROR
Test :func:`planemo.training.tutorial.get_hands_on_boxes_from_running_galaxy`. ... ERROR
Test :func:`planemo.training.tutorial.get_hands_on_boxes_from_local_galaxy`. ... ERROR
Test :func:`planemo.training.tutorial.get_hands_on_boxes_from_running_galaxy`. ... ERROR
```

Are you aware of any similar issue? I will investigate how the workflow and the extracted tools from these Galaxy instance look like (empty or not).

mblue9 commented:
@bebatut thanks for the file!!! That auto-generation is so cool 😎 🙌and going to save me a chunk of time (and tedium) 😄!!! 

I tried training_generate_from_wf without  `--galaxy_root` and I get the same error. I had run `training-init` without `--galaxy_root`.

jmchilton commented:
Fixed some linting issues and I added test decorators so the tests requiring Galaxy don't slow the "quick" version of the test suite. All the tests will still run with the full suite for both Python 2 and Python 3.

I couldn't love this anymore - it is absolutely amazing.
